x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -813,4 +854,99 @@ private static boolean isAsyncRelatedAction(String action) {
             || action.equals(EqlAsyncActionNames.EQL_ASYNC_GET_RESULT_ACTION_NAME)
             || action.equals(SqlAsyncActionNames.SQL_ASYNC_GET_RESULT_ACTION_NAME);
     }
+
+    /**
+     * A lazily loaded Set for authorized indices. It avoids loading the set if only contains check is required.
+     * It only loads the set if iterating through it is necessary, i.e. when expanding wildcards.
+     *
+     * NOTE that the lazy loading does not guarantee to run only once and is not meant to be used by multi-threads
+     * because loading multiple times can incur performance penalty (but not correctness).
+     */
+    private static class AuthorizedIndicesSet implements Set<String> {
+
+        private final Supplier<Set<String>> supplier;
+        private final Predicate<String> predicate;
+        private Set<String> authorizedIndices = null;
+
+        AuthorizedIndicesSet(Supplier<Set<String>> supplier, Predicate<String> predicate) {
+            this.supplier = Objects.requireNonNull(supplier);
+            this.predicate = Objects.requireNonNull(predicate);
+        }
+
+        private Set<String> getAuthorizedIndices() {
+            if (authorizedIndices == null) {
+                authorizedIndices = supplier.get();
+            }
+            return authorizedIndices;
+        }
+
+        @Override
+        public int size() {
+            return getAuthorizedIndices().size();
+        }
+
+        @Override
+        public boolean isEmpty() {
+            return getAuthorizedIndices().isEmpty();
+        }
+
+        @Override
+        public boolean contains(Object o) {
+            if (authorizedIndices == null) {
+                return predicate.test((String) o);
```

Comment:
Strictly speaking this violates the guarantees on the `Set` interface. You should return `false` if `o` is not a `String` (rather than throw CCE).
```suggestion
                if( o instanceof String s) {
                    return predicate.test(s);
                } else {
                    return false;
                }
```

Reply:
I thought about it. But [JDK doc](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Set.html#contains(java.lang.Object)) says it can throw `ClassCastException `. So I didn't do anything. Also adding `instanceof` check can in theory hide subtle usage error. So I am not sure if it helps with anything.